posthog/src/test/java/com/posthog/java/PostHogTest.java

```diff
@@ -56,7 +57,96 @@ public void testCaptureSimple() {
                         .isEqualTo(json.toString());
     }
 
-    // TODO: comprehensive public functions tests
+    @Test
+    public void testCaptureWithProperties() {
+        ph.capture("test id", "test event", new HashMap<String, Object>() {
+            {
+                put("movie_id", 123);
+                put("category", "romcom");
+            }
+        });
+        ph.shutdown();
+        assertEquals(1, sender.calls.size());
+        assertEquals(1, sender.calls.get(0).size());
+        JSONObject json = sender.calls.get(0).get(0);
+        assertThatJson("{\"distinct_id\":\"test id\",\"event\":\"test event\""
+                + ",\"properties\":{\"movie_id\":123,\"category\":\"romcom\"},\"timestamp\":\"" + instantExpected
+                + "\"}").isEqualTo(json.toString());
+    }
+
+    @Test
+    public void testIdentifySimple() {
+        ph.identify("test id", new HashMap<String, Object>() {
+            {
+                put("email", "john@doe.com");
+                put("proUser", false);
+            }
+        });
+        ph.shutdown();
+        assertEquals(1, sender.calls.size());
+        assertEquals(1, sender.calls.get(0).size());
+        JSONObject json = sender.calls.get(0).get(0);
+        assertThatJson("{\"distinct_id\":\"test id\",\"event\":\"$identify\""
+                + ",\"properties\":{\"$set\":{\"email\":\"john@doe.com\",\"proUser\":false}},\"timestamp\":\""
+                + instantExpected + "\"}").isEqualTo(json.toString());
+    }
+
+    @Test
+    public void testIdentifyWithSetOnce() {
+        ph.identify("test id", new HashMap<String, Object>() {
+            {
+                put("email", "john@doe.com");
+                put("proUser", false);
+            }
+        }, new HashMap<String, Object>() {
+            {
+                put("first_location", "colorado");
+                put("first_number", 5);
+            }
+        });
+        ph.shutdown();
+        assertEquals(1, sender.calls.size());
+        assertEquals(1, sender.calls.get(0).size());
+        JSONObject json = sender.calls.get(0).get(0);
+        assertThatJson("{\"distinct_id\":\"test id\",\"event\":\"$identify\""
+                + ",\"properties\":{\"$set\":{\"email\":\"john@doe.com\",\"proUser\":false}"
```

Comment:
i've been approving a few of these but certainly this is a nasty way to write tests.

Can we have a more structured variable and stringify that for testing?

Reply:
Good point on the pattern, though I did kinda want to make the test as dumb as possible & use what I could see being curl'ed directly to be sure I didn't mess up json parsing on both sides.